### ISSUE TITLE
Code File Use Cases

### ISSUE THREAD

**Issue by hexylena** [galaxyproject/galaxy#2714]
Please file your use cases for the code files tag which is long since deprecated, so we can find alternative solutions that are cleaner than "arbitrary code".

| Issue | Quick Description | Type |
| --- | --- | --- |
| #2713 | Multi-param validation | 1 |
| #2728 | `emboss_format_correctory.py` | 1 / has solution |
| #2729 | hardcoded list of options | 1 / has solution |
| #2712 | Dynamic options list (HTTP request) | 2 |
| #2725 | Dynamic options list (files in directory + API calls) | 2 |
| #2726 | Dynamic options list (files in directory) | 2 |
| #2727 | Dynamic options list (mysql) | 2 |
| #2883 | Dynamic options list (HTTP request / iRODS) | 2 |

**Comment by hrhotz** [comment]
Ours are:
#2725

**Comment by jmchilton** [comment]
I like @hrhotz's comment here https://github.com/galaxyproject/galaxy/issues/2712#issuecomment-237470961 and find it to be in the spirit of my own thoughts on this issue. A generic endpoint is useful to adapt Galaxy to infrastructure - but it should be generally _discouraged_ for public consumption tools (things in tools-iuc, things marked as "safe" in a tool shed, etc...).

So here is the trade off I would argue for - Galaxy continues indefinitely to provide such an extension point - but we are going to someday mark your tool as unsafe, we aren't going to give it a shiny gold star in the Tool Shed, we don't provide the same promises on backward compatibility (so the interface to Galaxy may change overtime, if you access Galaxy internals and we refactor - you need to modify your tools, things like that).

My proposal would be that we break these into two sets of use cases. The ones that we definitely want to solve with general purpose, declarative-ish mechanisms (Type 1) that are part of a spec, tested, documented, etc....  and ones where are willing to provide an extension point for so that people can produce local, non-idealized tools (Type 2).

I'd characterize them as follows:
- https://github.com/galaxyproject/galaxy/issues/2712 In general, I'd say dynamic options coming from outside sources are not reproducible and so are not idealized tools - I'd throw this in the Type 2 pile. If we can do things with metadata, etc... that would be great and maybe that would be enough to throw some of these into Type 1.
- https://github.com/galaxyproject/galaxy/issues/2713 This is Type 1 - there should be a way to do safe validation **in a sandbox**.
- https://github.com/galaxyproject/galaxy/issues/2725 Type 2 - it breaks Galaxy's model, we can't provide workflow support, history export, etc... but we should allow something like this - but it should be more thought of as a data source tool.
- https://github.com/galaxyproject/galaxy/issues/2883 Same as above - this is Type 2.
- https://github.com/galaxyproject/galaxy/issues/2729 This is largely solved by tool data tables - there should be better support for adapting different data sources, more documentation, ways to test data managers, etc... but all of these are side issues I think.
- https://github.com/galaxyproject/galaxy/issues/2728  https://github.com/galaxyproject/tools-devteam/blob/master/tools/emboss_5/emboss_format_corrector.py I think format actions cover this, the XSD has new and improved documentation on this. If we need to tweak it some we can - but I don't think code files are needed in general here.
- https://github.com/galaxyproject/galaxy/issues/2727 I'd argue Type 2 - this is a lot like https://github.com/galaxyproject/galaxy/issues/2712. I don't think there should be a high-level mechanism for this given the reproduciblity issues.

**Comment by jgoecks** [comment]
I agree with @jmchilton: keep the extension point, mark usage of it as a suboptimal practice, and enhance Galaxy to mitigate its use.

**Comment by hexylena** [comment]
I've re-sorted the list, "type 1" sound like they're basically all handled except for #2713, multi-param validation. I'll close those when I find time to write a good response pointing people who find those issues to the right location for how we should be doing this now.
